### PR TITLE
iteration, bulk appendix

### PR DESCRIPTION
**Pull request by andre-merzky**

### CODE REVIEW

File: specification.md [hunk 1]
```diff
@@ -932,13 +825,14 @@ JobState getState()
...
-implementations have the discretion of implementing a relevant
-`Timestamp` class.
+<a name="jobstatus-gettime"></a>
+```java
+Timestamp getTime()
+```
+Returns the time at which the job has entered this state. The `Timestamp` class
+is expected to be provided by the standard library of the language in which the
+library is implemented. If such a class is not provided, implementations have
+the discretion of implementing a relevant `Timestamp` class.
```
Comment (andre-merzky) [inline specification.md:835]: TODO AM: msg: clarify purpose: reason for state transitions, error messages for FAILED state, etc.

File: specification.md [hunk 2]
```diff
@@ -200,14 +208,19 @@ exascale machines, as well as propose such an API. A job management API
...
-job management library, through its  API,  is invoked by a client
+job management library, through its API, is invoked by a client
 application.
 
 Traditionally, job management is implemented on supercomputers by Local
 Resource Managers (LRMs), such as PBS/Torque, SLURM, etc. In a first
 approximation, a job management API is understood as an abstraction layer
 on top of various LRMs.
 
+Non-Traditionally, job management is also provided by execution managers, which
```
Comment (SteVwonder) [inline specification.md:219]:
```suggestion
Nontraditionally, job management is also provided by execution managers, which
```

[[1]](https://www.merriam-webster.com/dictionary/nontraditional#other-words)

File: specification.md [hunk 3]
```diff
@@ -398,10 +417,10 @@ methods.
 
 ## Interaction with LRMs and Scalability
 
-Implementations must use bulk status operations when interacting with
+Implementations should use bulk status operations when interacting with
```
[the remark this answers is not in the document]
Comment (SteVwonder) [reply specification.md:420]: Maybe something like:

```suggestion
Implementations must use bulk status operations, if available, when interacting with
```

File: specification.md [hunk 4]
```diff
@@ -410,23 +429,22 @@ extract the information about the relevant jobs from the result.
...
-System 2 receives the message "B" before "A". In terms of jobs,
-this may make it appear as if seemingly impossible things are happening,
-such as a job starting to run after it has completed. Implementations
-must ensure that client code does not receive events in orders that are
-clearly impossible. The specifics of how this must be handled by
-implementations is detailed in [`Job.getStatus()`](#job-getstatus) and
-[`JobState`](#jobstate).
+TODO AM: state model should be defined before this section.
 
+The API specification pre-scribes a state model for jobs.  Backend
```
Comment (SteVwonder) [inline specification.md:434]:
```suggestion
The API specification prescribes a state model for jobs.  Backend
```

File: specification.md [hunk 5]
```diff
@@ -949,12 +843,16 @@ int? getExitCode()
...
-<a name="jobstatus-getmessage"></a>
+<a name="jobstatus-getcontext"></a>
 ```java
-String? getMessage()
+Map<String, Any>? getContext()
 ```
 
-Returns the message associated with this status, if any.
+Returns additional, backend specific meta data associated with this status, if
+any.  Those meta data may include details on the state transition, backend
```
Comment (SteVwonder) [inline specification.md:852]:
```suggestion
any.  Those metadata may include details on the state transition, backend
```

File: specification.md [hunk 6]
```diff
@@ -166,14 +173,15 @@ exec/popen.
...
-    beyond the scope of this API
+    beyond the scope of this (version of the) API
 
     - We would like the interface to be general and applicable to
     commonly deployed LRMs, cloud systems, etc. (I know some of this is
     said above)
 
     - We are focused on executing a process (e.g., popen rather than
     function call)
+    AM: Why actually?  The only point where this shows is the jobspec.
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply specification.md:184]: Yeah, ignore that for now, will remove.  I was thinking out loud here: I wanted to check if one could, for example, add a `FunctionSpecification` later on to trigger a   function execution on the backend, for example.

File: specification.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (andre-merzky) [reply specification.md:420]: I am not a fan of forbidding stupid implementations, really.  I mean, semantically that submit loop is correct, so who am I to tell a developer that he must be efficient - that should not be an API compliance issue?

File: specification.md [hunk 7]
```diff
@@ -226,8 +239,14 @@ pseudo-code which almost surely will require modifications to be usable.
 
 ## Layers
 
+AM: After the different discussions we had, I am not sure if the API should
+really distinguish these layers.  The *implementation* very much needs to, but
+I don't see how the description of a job, the management of its execution, or
+the semantics of its state would differ between those layers.  I would thus
+suggest to limit this discussion to the implementation of the API.
+
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply specification.md:247]: Yes, I'll remove that text - will open an ticket about that point.

File: specification.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (hategan) [reply specification.md:420]: My question is what's the point of having bulk submit if we don't heavily nudge implementers in the direction of making it more efficient than the loop. It being semantically correct seems like the very reason why we should be more forceful here in making it clear that there's more than blindly correct semantics that is expected from implementations.

File: specification.md [hunk 8]
```diff
@@ -153,6 +155,11 @@ any "how do you do x?", please add here)
 exec/popen.
 
 - [ ] add some text about pilot jobs / reslicing
+      AM: IMHO, the easiest way to represent such hierarchies in the API is to
+      be able to obtain a JobExecutor from a (pilot) Job.  Alternatively,
+      a pilot job entering ACTIVE state can provide a JobExecutor endpoint as
+      metda data - but then we would require the non-standardized meta data to
+      be used.
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply specification.md:162]: discussion moved here: #20

File: specification.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (andre-merzky) [reply specification.md:406]: Implementers are usually interested in making their stuff useful, so they usually have an incentive to make things fast, stable, scalable, documented, ...  I am just saying that an API spec should not be normative about those things - an API spec should say: if method A is called with parameters C and D, then action E is happening unless error F occurs, causing an exception G, ,,, and so on.  It should *not* say: an implementation must be efficient and well documented and scalable etc. 

I don't want to get hung up on that single should / must, I can revert that.  It's more of a general point: IMHO an API spec defines syntax and semantics of an API, but should not put any constraints on the implementation.  Is an implementation which is slow not compliant?  I implementation A with bulk RPC is slower than implementation B which calls submit individually, but concurrently?  How would you even tell for a closed-source implementation, and ascertain compliance?

**Comment by andre-merzky** [comment]
This PR is obsolete by now.

**Pull request closed without merging**